File: backend/app/schemas/project_wizard.py
"""Schemas for the end-to-end project creation wizard."""

from __future__ import annotations

import uuid
from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.ai import ProjectConfig


class PaperUploadResponse(BaseModel):
    """Parsed paper details and the newly created draft project."""

    project_id: uuid.UUID
    project_name: str
    paper_title: str
    abstract: str
    authors: list[str] = Field(default_factory=list)
    keywords: list[str] = Field(default_factory=list)


class ExistingAssetsProjectRequest(BaseModel):
    """Create a project that starts from prepared code and data."""

    name: str = Field(min_length=1, max_length=200)


class ExistingAssetsProjectResponse(BaseModel):
    """Draft project created for the existing-assets workflow."""

    project_id: uuid.UUID
    project_name: str
    workflow: Literal["existing_assets"] = "existing_assets"


class PaperAnalysisResponse(BaseModel):
    """Structured implementation-oriented reading of the source paper."""

    summary: str
    research_problem: str
    method_steps: list[str] = Field(default_factory=list)
    datasets: list[str] = Field(default_factory=list)
    metrics: list[str] = Field(default_factory=list)
    implementation_requirements: list[str] = Field(default_factory=list)
    compute_requirements: list[str] = Field(default_factory=list)
    reproducibility_risks: list[str] = Field(default_factory=list)
    model: str


class DataSelectionResponse(BaseModel):
    """Manifest for a selected file or folder on the SSH server."""

    ready: bool
    source: Literal["remote"] = "remote"
    kind: Literal["file", "folder"]
    selected_name: str
    path: str
    file_count: int
    total_bytes: int


class RemoteDataEntryResponse(BaseModel):
    """One selectable data entry returned by the remote browser."""

    name: str
    path: str
    kind: Literal["file", "folder"]
    size: int


class RemoteDataListingResponse(BaseModel):
    """A bounded directory listing from the selected SSH server."""

    current_path: str
    parent_path: str | None = None
    entries: list[RemoteDataEntryResponse] = Field(default_factory=list)
    truncated: bool = False


class RemoteDataSelectionRequest(BaseModel):
    """One remote file or folder chosen through the server browser."""

    path: str = Field(min_length=1, max_length=4096)
    kind: Literal["file", "folder"]


class RemoteCodeImportRequest(BaseModel):
    """Remote code directory and its relative training entrypoint."""

    path: str = Field(min_length=1, max_length=4096)
    entrypoint: str = Field(min_length=1, max_length=500)
    arguments: str = Field(default="", max_length=2000)


class RemoteCodeImportResponse(BaseModel):
    """Git-backed snapshot imported from an SSH server directory."""

    ready: bool
    source: Literal["remote"] = "remote"
    selected_name: str
    path: str
    entrypoint: str
    arguments: list[str] = Field(default_factory=list)
    file_count: int
    total_bytes: int
    skipped_count: int


class SshConnectionResponse(BaseModel):
    """Verified remote execution target without its credentials."""

    ready: bool
    host: str
    port: int
    username: str
    auth_type: Literal["password", "key"]
    host_key_fingerprint: str
    capabilities: dict[str, str | bool]


class PreparationStatusResponse(BaseModel):
    """Current hard-gate status for an experiment project."""

    workflow: Literal["paper_reproduction", "existing_assets"]
    api_key_ready: bool
    paper_analysis_ready: bool
    data_ready: bool
    execution_ready: bool
    code_ready: bool
    ready_to_generate: bool
    ready_to_start: bool
    ai_profile_id: uuid.UUID | None = None
    ssh_profile_id: uuid.UUID | None = None
    data: DataSelectionResponse | None = None
    code: RemoteCodeImportResponse | None = None
    execution: SshConnectionResponse | None = None
    missing: list[str] = Field(default_factory=list)


class DialogAnswerRequest(BaseModel):
    """One answer in a guided configuration dialog."""

    session_id: str
    answer: str = Field(min_length=1, max_length=2000)


class DialogResponse(BaseModel):
    """Either the next question or a completed project configuration."""

    session_id: str
    complete: bool = False
    question: str | None = None
    options: list[str] = Field(default_factory=list)
    input_type: Literal["single", "multi", "text"] | None = None
    config: ProjectConfig | None = None


class GeneratedFile(BaseModel):
    """One generated source file exposed for review."""

    path: str
    language: str
    content: str


class CodeGenerationResponse(BaseModel):
    """Generated framework and the agent's explanation."""

    project_id: uuid.UUID
    files: list[GeneratedFile]
    summary: str


class WizardProjectSnapshot(BaseModel):
    """Persisted wizard state used to resume an unfinished project."""

    project_id: uuid.UUID
    project_name: str
    paper_title: str
    abstract: str
    authors: list[str] = Field(default_factory=list)
    keywords: list[str] = Field(default_factory=list)
    analysis: PaperAnalysisResponse | None = None
    dialog_complete: bool = False
    config: ProjectConfig | None = None
    preparation: PreparationStatusResponse
    files: list[GeneratedFile] = Field(default_factory=list)


class SaveCodeRequest(BaseModel):
    """Reviewed files to persist in the project repository."""

    files: list[GeneratedFile] = Field(min_length=1, max_length=50)


class SaveCodeResponse(BaseModel):
    """Commit created after saving reviewed source files."""

    project_id: uuid.UUID
    commit_sha: str
    files_saved: int


class StartExperimentResponse(BaseModel):
    """Initial experiment queued from the completed wizard."""

    project_id: uuid.UUID
    experiment_id: uuid.UUID
    status: Literal["queued"]

File: frontend/src/services/projectWizard.ts
import axios from "axios";
import api from "./api";

export interface PaperUploadResult {
  project_id: string;
  project_name: string;
  paper_title: string;
  abstract: string;
  authors: string[];
  keywords: string[];
}

export interface PaperAnalysis {
  summary: string;
  research_problem: string;
  method_steps: string[];
  datasets: string[];
  metrics: string[];
  implementation_requirements: string[];
  compute_requirements: string[];
  reproducibility_risks: string[];
  model: string;
}

export interface DataSelection {
  ready: boolean;
  source: "remote";
  kind: "file" | "folder";
  selected_name: string;
  path: string;
  file_count: number;
  total_bytes: number;
}

export interface RemoteDataEntry {
  name: string;
  path: string;
  kind: "file" | "folder";
  size: number;
}

export interface RemoteDataListing {
  current_path: string;
  parent_path: string | null;
  entries: RemoteDataEntry[];
  truncated: boolean;
}

export interface SshConnection {
  ready: boolean;
  host: string;
  port: number;
  username: string;
  auth_type: "password" | "key";
  host_key_fingerprint: string;
  capabilities: Record<string, string | boolean>;
}

export interface PreparationStatus {
  workflow: "paper_reproduction" | "existing_assets";
  api_key_ready: boolean;
  paper_analysis_ready: boolean;
  data_ready: boolean;
  execution_ready: boolean;
  code_ready: boolean;
  ready_to_generate: boolean;
  ready_to_start: boolean;
  ai_profile_id: string | null;
  ssh_profile_id: string | null;
  data: DataSelection | null;
  code: RemoteCodeImport | null;
  execution: SshConnection | null;
  missing: string[];
}

export interface ExistingAssetsProject {
  project_id: string;
  project_name: string;
  workflow: "existing_assets";
}

export interface RemoteCodeImport {
  ready: boolean;
  source: "remote";
  selected_name: string;
  path: string;
  entrypoint: string;
  arguments: string[];
  file_count: number;
  total_bytes: number;
  skipped_count: number;
}

export interface ProjectConfig {
  improvement_targets: string[];
  target_metrics: Record<string, number>;
  max_iterations: number;
  summary: string;
}

export interface DialogResult {
  session_id: string;
  complete: boolean;
  question?: string;
  options: string[];
  input_type?: "single" | "multi" | "text";
  config?: ProjectConfig;
}

export interface GeneratedFile {
  path: string;
  language: string;
  content: string;
}

export interface CodeGenerationResult {
  project_id: string;
  files: GeneratedFile[];
  summary: string;
}

export interface WizardProjectSnapshot {
  project_id: string;
  project_name: string;
  paper_title: string;
  abstract: string;
  authors: string[];
  keywords: string[];
  analysis: PaperAnalysis | null;
  dialog_complete: boolean;
  config: ProjectConfig | null;
  preparation: PreparationStatus;
  files: GeneratedFile[];
}

export interface StartExperimentResult {
  project_id: string;
  experiment_id: string;
  status: "queued";
}

export const projectWizardApi = {
  async createExistingAssetsProject(
    name: string,
  ): Promise<ExistingAssetsProject> {
    const response = await api.post<ExistingAssetsProject>(
      "/api/projects/wizard/existing-assets",
      { name },
    );
    return response.data;
  },

  async uploadPaper(
    file: File,
    projectName: string,
  ): Promise<PaperUploadResult> {
    const formData = new FormData();
    formData.append("paper", file);
    formData.append("project_name", projectName);
    const response = await api.post<PaperUploadResult>(
      "/api/projects/wizard/upload",
      formData,
      {
        headers: { "Content-Type": "multipart/form-data" },
        timeout: 60_000,
      },
    );
    return response.data;
  },

  async startDialog(projectId: string): Promise<DialogResult> {
    const response = await api.post<DialogResult>(
      `/api/projects/${projectId}/wizard/dialog/start`,
    );
    return response.data;
  },

  async selectConfigurations(
    projectId: string,
    selection: {
      ai_profile_id?: string | null;
      ssh_profile_id?: string | null;
    },
  ): Promise<PreparationStatus> {
    const response = await api.put<PreparationStatus>(
      `/api/projects/${projectId}/wizard/configurations`,
      selection,
    );
    return response.data;
  },

  async analyzePaper(projectId: string): Promise<PaperAnalysis> {
    const response = await api.post<PaperAnalysis>(
      `/api/projects/${projectId}/wizard/analyze`,
      undefined,
      { timeout: 180_000 },
    );
    return response.data;
  },

  async browseRemoteData(
    projectId: string,
    path?: string,
  ): Promise<RemoteDataListing> {
    const response = await api.get<RemoteDataListing>(
      `/api/projects/${projectId}/wizard/remote-data`,
      {
        params: path ? { path } : undefined,
        timeout: 45_000,
      },
    );
    return response.data;
  },

  async selectRemoteData(
    projectId: string,
    entry: { path: string; kind: "file" | "folder" },
  ): Promise<DataSelection> {
    const response = await api.put<DataSelection>(
      `/api/projects/${projectId}/wizard/remote-data`,
      entry,
      {
        timeout: 45_000,
      },
    );
    return response.data;
  },

  async importRemoteCode(
    projectId: string,
    input: { path: string; entrypoint: string; arguments: string },
  ): Promise<RemoteCodeImport> {
    const response = await api.post<RemoteCodeImport>(
      `/api/projects/${projectId}/wizard/remote-code`,
      input,
      { timeout: 300_000 },
    );
    return response.data;
  },

  async getPreparation(projectId: string): Promise<PreparationStatus> {
    const response = await api.get<PreparationStatus>(
      `/api/projects/${projectId}/wizard/preparation`,
    );
    return response.data;
  },

  async getSnapshot(projectId: string): Promise<WizardProjectSnapshot> {
    const response = await api.get<WizardProjectSnapshot>(
      `/api/projects/${projectId}/wizard/snapshot`,
    );
    return response.data;
  },

  async answerDialog(
    projectId: string,
    sessionId: string,
    answer: string,
  ): Promise<DialogResult> {
    const response = await api.post<DialogResult>(
      `/api/projects/${projectId}/wizard/dialog/answer`,
      { session_id: sessionId, answer },
    );
    return response.data;
  },

  async generateCode(projectId: string): Promise<CodeGenerationResult> {
    const response = await api.post<CodeGenerationResult>(
      `/api/projects/${projectId}/wizard/generate`,
      undefined,
      // Framework generation can span many model tool-use turns. The model
      // provider remains responsible for its own request-level timeouts.
      { timeout: 0 },
    );
    return response.data;
  },

  async saveCode(projectId: string, files: GeneratedFile[]): Promise<void> {
    await api.put(`/api/projects/${projectId}/wizard/code`, { files });
  },

  async startExperiment(projectId: string): Promise<StartExperimentResult> {
    const response = await api.post<StartExperimentResult>(
      `/api/projects/${projectId}/wizard/start`,
    );
    return response.data;
  },
};

export function getWizardError(error: unknown): string {
  if (axios.isAxiosError(error)) {
    const detail = error.response?.data?.detail;
    if (typeof detail === "string") {
      return detail;
    }
    if (detail && typeof detail.detail === "string") {
      const missing = Array.isArray(detail.missing)
        ? `：${detail.missing.join("、")}`
        : "";
      return `${detail.detail}${missing}`;
    }
  }
  return error instanceof Error ? error.message : "操作未完成，请重试。";
}
